zp-relayer/workers/sentTxWorker.ts

```diff
@@ -124,81 +80,92 @@ export async function createSentTxWorker<T extends EstimationType>(gasPrice: Gas
+        }
+        logger.info('Marking ids %j as failed', waitingJobIds)
+        await markFailed(redis, waitingJobIds)
+        logger.info('%s Rescheduling %d jobs to process...', logPrefix, waitingJobs.length)
+        const rescheduledIds = await Promise.all(reschedulePromises)
+
+        return [SentTxState.REVERT, txHash, rescheduledIds] as SentTxResult
       }
     } else {
       // Resend with updated gas price
       const txConfig = job.data.txConfig
 
       const oldGasPrice = job.data.gasPriceOptions
-      const fetchedGasPrice = gasPrice.getPrice()
 
-      const newGasPrice = chooseGasPriceOptions(oldGasPrice, fetchedGasPrice)
+      const fetchedGasPrice = await gasPrice.fetchOnce()
+      const oldWithExtra = addExtraGasPrice(oldGasPrice, config.minGasPriceBumpFactor, null)
+      const newWithExtra = addExtraGasPrice(fetchedGasPrice, config.gasPriceSurplus, null)
 
-      logger.warn('Tx %s is not mined; updating gasPrice: %o -> %o', txHash, oldGasPrice, newGasPrice)
+      const newGasPrice = chooseGasPriceOptions(oldWithExtra, newWithExtra)
+
+      logger.warn('%s Tx %s is not mined; updating gasPrice: %o -> %o', logPrefix, txHash, oldGasPrice, newGasPrice)
 
       const newTxConfig = {
         ...txConfig,
         ...newGasPrice,
       }
 
       try {
-        if (txType === TxType.PERMITTABLE_DEPOSIT) {
-          const deadline = (txData as PermittableDepositTxData).deadline
-          await checkAssertion(() => checkDeadline(toBN(deadline), config.permitDeadlineThresholdResend))
-        }
-
         const newTxHash = await signAndSend(newTxConfig, config.relayerPrivateKey, web3)
-
-        // Add updated job
-        await sentTxQueue.add(
-          newTxHash,
-          {
-            ...job.data,
-            txHash: newTxHash,
-            txConfig: newTxConfig,
-            gasPriceOptions: newGasPrice,
-          },
-          {
-            priority: txConfig.nonce,
-            delay: config.sentTxDelay,
-          }
-        )
-        return [SentTxState.RESEND, newTxHash] as SentTxResult
+        // Update job
+        await job.update({
+          ...job.data,
+          txHash: newTxHash,
+          txConfig: newTxConfig,
+          gasPriceOptions: newGasPrice,
+        })
+        await job.updateProgress({ txHash: newTxHash, gasPrice: newGasPrice })
       } catch (e) {
         const err = e as Error
-        logger.warn('%s: Tx resend failed for %s: %s', logPrefix, txHash, err.message)
-        if (isSameTransactionError(err) || isGasPriceError(err)) {
-          // Force update gas price
-          gasPrice.start()
-        } else if (isNonceError(err)) {
+        logger.warn('%s Tx resend failed for %s: %s', logPrefix, txHash, err.message)
+        // TODO: Should we handle other tx sending errors here?
+        if (isNonceError(err)) {
           await updateNonce(await readNonce(true))
```

Comment:
As we now strictly relate each job sending attempt (including resends, excluding re-queueing after rollback) with a particular nonce, forced nonce updates will mess up everything in our logic.

Let's find a way of getting rid of it in both workers.

Reply:
I think in current implementation we can safely get rid of handling nonce errors, but still we should have a recovery mechanism in some place in case such error is thrown (whatever the reason was) to prevent a complete deadlock